File: 503. Next Greater Element II.java
/**
 * Loop twice to find out the circular NGE
 * Time O(n)
 * Spcae O(n)
 */
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int n = nums.length;
        int[] res = new int[n];
        Arrays.fill(res, -1);
        Deque<Integer> stack = new ArrayDeque<>();
        for (int i=0; i<n*2-1; i++) {
            while (!stack.isEmpty() && nums[stack.peek()] < nums[i%n]) {
                int temp = stack.pop();
                res[temp] = nums[i%n];
            }
            stack.push(i%n);
        }
        return res;
    }
}

File: 503. Next Greater Element II.py
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        if not nums:
            return []
        n = len(nums)
        res = [-1] * n
        stack = []
        for i in range(n):
            if not stack:
                stack.append(i)
            else:
                while len(stack) > 0 and nums[stack[-1]] < nums[i]:
                    res[stack.pop()] = nums[i]
                stack.append(i)
        for i in range(n):
            while len(stack) > 0 and nums[stack[-1]] < nums[i]:
                res[stack.pop()] = nums[i]
        return res